### DISCUSSION TITLE
In Shadow, how do I generate traffic of a particular website.

### DISCUSSION THREAD

**Discussion by beenhuang** [shadow/shadow#1714] [General]
In Shadow, how do I generate traffic of a particular website, such as www.facebook.com, www.google.com, etc.

**Comment by sporksmith** [comment]
It depends what you're trying to accomplish, but there's no easy answer. What are you trying to learn from your simulations?

You won't be able to perfectly emulate the site's dynamic handling, since you don't have access to their binaries to run under shadow. You could probably scrape some of their content and set up a static web server to serve that content, but there are a lot of corner cases. Using a reverse proxy cache like squid to build up an initial cache outside of shadow, and then serve it under shadow, might work.

**Reply by beenhuang** [reply]
thanks for your reply,  I want to capture the cell sequence of a particular website using Tor in the  Shadow environment. How do I get users using Tor inside Shadow to access reverse proxy cache outside Shadow?

**Reply by sporksmith** [reply]
> How do I get users using Tor inside Shadow to access reverse proxy cache outside Shadow?

I was thinking you could first run squid natively to build up a cache on disk, and then run it under shadow in your simulation to serve that cache.

For simpler sites, using something like `wget` to crawl a site, and then a simple web server like `ngingx` under Shadow could be simpler.

To be clear, this isn't a supported or tested use-case. I'm trying to give some ideas of how it could work, but we can't guarantee that it *will* work :). There are lots of subtleties around trying to mirror and re-serve a web site like this in general, which I'm happy to discuss but are a bit orthogonal to Shadow itself.

Running new software under shadow (like nginx or squid) is also likely to uncover bugs and missing functionality in Shadow, but generally we're happy to file and fix such bugs :)
